[Version 9.0] Feature support for native sized integers (`nint`, and `nuint`) (#1457)
dotnet/csharpstandard: standard/types.md

### 8.3.5 Simple types

> C# provides a set of predefined `struct` types called the simple types. The simple types are identified through keywords, but these keywords are simply aliases for predefined `struct` types in the `System` namespace, as described in the table below.

Reviewer: I struck the first sentence as this is already stated in 8.3.1.

  `bool`    |   `System.Boolean`
  `decimal` |   `System.Decimal`

Every simple type has members. Each simple type that is an alias for a predefined struct type, has that struct type’s members.

> > *Example*: `int` has the members declared in `System.Int32` and the members inherited from `System.Object`, and the following statements are permitted:

Reviewer 2: This is possibly misleading, the Standard requires no members on `System.Int32`, but does require a few on `System.Object`. Maybe it should be something like:

```suggestion
Every simple type has members. Each simple type that is an alias for a predefined struct type, has that struct type’s members.

> *Example*: `int` has any implementation-specific members declared in `System.Int32` and the required and implementation-specific members inherited from `System.Object`, and the following statements are permitted:
```

But is something like this *too* accurate?

Author: I made some slight word choice changes here.

>
> *end note*.

Although `nint` and `nuint` shall be represented by the types `System.IntPtr` and `System.UIntPtr`, respectively, `nint` and `nuint` are *not* aliases for those types. As such, not all members of the corresponding `System` types are defined for `nint` and `nuint`. Instead, the compiler shall make available additional conversions and operations for the types `System.IntPtr` and `System.UIntPtr` when used in the context of native integer types.

Consider the following:

> 

Reviewer: ```suggestion
```

>
> *end note*.

Although `nint` and `nuint` shall be represented by the types `System.IntPtr` and `System.UIntPtr`, respectively, `nint` and `nuint` are *not* aliases for those types. As such, not all members of the corresponding `System` types are defined for `nint` and `nuint`. Instead, the compiler shall make available additional conversions and operations for the types `System.IntPtr` and `System.UIntPtr` when used in the context of native integer types.

Consider the following:

<!-- Example: {template:"standalone-console-without-using", name:"SimpleTypes2", expectedErrors:["CS0266"], ignoredWarnings:["CS0219"]} -->
```csharp
nint a1 = 1;           // OK
System.IntPtr a2 = 1;  // Error: no implicit conversion

> ```

Reviewer: ```suggestion
```

> While the implementation provides operations and conversions for `nint` and `nuint` that are appropriate for integer types, those operations and conversions are not available on the `System` type counterparts. Similarly,

Reviewer: ```suggestion
While the implementation provides operations and conversions for `nint` and `nuint` that are appropriate for integer types, those operations and conversions are not available on the `System` type counterparts. For example,
```

> *Note*: Operations performed through `dynamic` binding on `System.IntPtr` and `System.UIntPtr` values do not have access to the `nint` and `nuint` operators. *end note*

`sizeof(nint)` and `sizeof(nuint)` return the size of a native integer ([§23.6.9](unsafe-code.md#2369-the-sizeof-operator)).

`typeof(nint)` is `typeof(System.IntPtr)`, and `typeof(nuint)` is `typeof(System.UIntPtr)`.

<!-- C# 11: The following three paragraphs about overriding/hiding equivalence can be removed
     when nint/nuint become true aliases for IntPtr/UIntPtr. -->

`nint` and `System.IntPtr`, and `nuint` and `System.UIntPtr`, are considered equivalent for overriding, hiding, and implementing.

Overloads cannot differ by `nint` and `System.IntPtr`, or by `nuint` and `System.UIntPtr`, alone. However, overrides and implementations may differ by `nint` and `System.IntPtr`, or by `nuint` and `System.UIntPtr`, alone.

Methods hide other methods that differ by `nint` and `System.IntPtr`, or by `nuint` and `System.UIntPtr`, alone.

> 

Reviewer 2: The difference between being *represented by* and being *an alias of* turn out to be very minimal and a lot of this section is not required. By just defining these differences we get:

```suggestion
<!-- C# 11: In C# 11, nint and nuint become true aliases for System.IntPtr and System.UIntPtr. The following paragraphs describing the non-alias relationship should be updated or removed. -->

The types `nint` and `nuint` are represented by the types `System.IntPtr` and `System.UIntPtr`, respectively, and are *not* aliases for these types. In this context being *represented by* means:

- The only members directly accessible for `nint` and `nuint` are the required methods of `Object` ([§C.2](standard-library.md#c2-standard-library-types-defined-in-isoiec-23271)). Any other members of System.IntPtr` and `System.UIntPtr` may be accessed via those types.
- Operations performed through `dynamic` binding on `System.IntPtr` and `System.UIntPtr` values do not have access to the `nint` and `nuint` operators.
- In all other respects `nint` and `nuint` behave as if they are aliases of `System.IntPtr` and `System.UIntPtr`.

```

### 8.3.6 Integral types

C# supports the following integral types, with the sizes and value ranges, as shown:

- The `sbyte` type represents signed 8-bit integers with values from `-128` to `127`, inclusive.
- The `byte` type represents unsigned 8-bit integers with values from `0` to `255`, inclusive.
- The `short` type represents signed 16-bit integers with values from `-32768` to `32767`, inclusive.
- The `ushort` type represents unsigned 16-bit integers with values from `0` to `65535`, inclusive.
- The `int` type represents signed 32-bit integers with values from `-2147483648` to `2147483647`, inclusive.
- The `uint` type represents unsigned 32-bit integers with values from `0` to `4294967295`, inclusive.
- The `nint` type represents a ***native signed integer*** whose size and value range are implementation-defined, but which shall be either that of `int` or `long`.
- The `nuint` type represents a ***native unsigned integer*** whose size and value range are implementation-defined, but which shall be either that of `uint` or `ulong`. The size of a native unsigned integer shall be the same as that of a native signed integer.

>   > *Note*: Unlike the other integral types `nint` and `nuint` do not have `const` fields called `MinValue` and `MaxValue`. As a practical consideration, implementations are expected to be at least 32-bit, so the value range of `nint` includes at least that of `int`, and the value range of `nuint` includes at least that of `uint`. *end note*

Reviewer 2: `MinValue` & `MaxValue` are not part of the Standard, implementations also require `nint` & `nuint` to be 32-bit or 64-bit. Just delete the *Note*.

```suggestion
 ```

>
> *end note*.

> Although `nint` and `nuint` shall be represented by the types `System.IntPtr` and `System.UIntPtr`, respectively, `nint` and `nuint` are *not* aliases for those types. As such, not all members of the corresponding `System` types are defined for `nint` and `nuint`. Instead, the compiler shall make available additional conversions, unary operators (§12.9) and arithmetic operators (§12.12) for the types `System.IntPtr` and `System.UIntPtr` when used in the context of native integer types.

Reviewer 2: As covered in this [comment](https://github.com/dotnet/csharpstandard/pull/1457/files#r2743902094) this seems to be largely incorrect from a Standard viewpoint. The only methods hat are required by the Standard are `Object`’s `Equals`, `GetHashCode`, `GetType` and `ToString`.

So this might be something like:

```suggestion
Although `nint` and `nuint` shall be represented by the types `System.IntPtr` and `System.UIntPtr`, respectively, `nint` and `nuint` are *not* aliases for those types. As such, the only members directly accessible for `nint` and `nuint` are the required methods of `Object` (§insert).
```

That, at minimum, needs wordsmithing!

Author: With the reworking of this section, do you want this added, or does the new version address your concerns here?

> While the language provides operations and conversions for `nint` and `nuint` that are appropriate for integer types, those operations and conversions are not available on the `System` type counterparts. For example,

Reviewer 2: These few lines give rise to a few questions:

> Although `nint` and `nuint` shall be represented by the types `System.IntPtr` and `System.UIntPtr`, respectively, `nint` and `nuint` are *not* aliases for those types. As such, not all members of the corresponding `System` types are defined for `nint` and `nuint`.

This seems to say there are members defined on `IntPtr` & `UIntPtr` that are explicitly not accessible via objects/values of type `nint` & `nuint`.

So is the deviation of `nint` & `nuint` from the other numeric types the restriction that not all members of the `System.X` type used for them may be accessed whether those members?

Remember under the conformance rules an implementation may provide types with additional members which are then accessible, are these native types avoiding that by not being “aliases“?

> Instead, the compiler shall make available additional conversions, unary operators (§12.9) and arithmetic operators (§12.12) for the types `System.IntPtr` and `System.UIntPtr` when used in the context of native integer types.

First, and obviously, the “compiler” shall do not such thing – the *language* defines the operators, e.g. §12.10.5 (v7):

> The predefined addition operators are listed below. For numeric and enumeration types, the predefined addition operators compute the sum of the two operands.

Where in the Standard does it ever state that these predefined operators are members of the “aliased” `System.X` type?

What difference is being asserted here that the “compiler” is making available that the language doesn’t for the other numeric types?

From a language perspective there are very few required members of any of numeric types. Take a look at `System.Int` and in §C no members are listed, look at the adopted spec from the CLI and it contains: 2 fields (min & max values), 2 compare methods, 2 equality methods, 1 hash code method, 4 parsing methods, and 4 conversions to string. What isn’t there is more notable than what is, e.g. not a single operator.

> While the language provides operations and conversions for `nint` and `nuint` that are appropriate for integer types, those operations and conversions are not available on the `System` type counterparts.

So how does that differ from `System.Int` above, it also has no operators…

Now today .NET’s `System.Int` does, for example, implement `IAdditionOperators<TSelf,TOther,TResult>` which provides and addition operator but it is not in the C# Standard and use Roslyn to add to integers and it does not call it – nothing has changed, the language defines the operators, the implementation implements them using the features of its target architecture (for Roslyn compiling C#9 the CLR’s `add` for both `int` and `nint`).

So what is *different* about `nint` and `nuint` that needs to go into the Standard (v9 & v11)? At present this PR has no native int operators listed in §12.10.5 (v7) which seems rather key, but does have stuff about the compiler making additional things available which seems entirely irrelevant.

Looking at the source material linked at the top of the PR it is clear it is heavily implementation specific, going as far as listing CLR instructions.

All the implementation specific stuff picked up from the source material needs to be elided, and the core of the C# languages changes found in what is left – and once that is done I suspect the differences between C#9 & C#11 will be much reduced.

<!-- Example: {template:"code-in-main-without-using", name:"SimpleTypes3", expectedException:"RuntimeBinderException"} -->
```csharp
M((nint)1);

static void M(dynamic d)
{
    var v = d >> 2; // RuntimeBinderException: '>>' cannot be applied to operands
                    // of type System.IntPtr/System.UIntPtr and int
}
```

> The only constructor for `nint` or `nuint` is the parameter-less constructor.

Reviewer 2: Does the Standard ever mention the constructor for `int` and the other numeric types? I suspect this in an implementation-specific detail.

```suggestion
```

The following members of `System.IntPtr` and `System.UIntPtr` are explicitly excluded from `nint` or `nuint`:

```csharp
// constructors
// arithmetic operators
// implicit and explicit conversions
public static readonly IntPtr Zero; // use 0 instead
public static int Size { get; }     // use sizeof() instead
public static IntPtr Add(IntPtr pointer, int offset);
public static IntPtr Subtract(IntPtr pointer, int offset);
public int ToInt32();
public long ToInt64();
public void* ToPointer();
```

The remaining members of `System.IntPtr` and `System.UIntPtr` are implicitly included in `nint` and `nuint`.

> Interfaces implemented by `System.IntPtr` and `System.UIntPtr` are implicitly included in `nint` and `nuint`, with occurrences of the underlying types replaced by the corresponding native integer types. For example, if `IntPtr` implements `ISerializable`, `IEquatable<IntPtr>`, and `IComparable<IntPtr>`, then `nint` implements `ISerializable`, `IEquatable<nint>`, and `IComparable<nint>`.

Reviewer 2: See this [comment](https://github.com/dotnet/csharpstandard/pull/1457/files#r2752405643) above. All the mentioned methods are implementation-specific.

Line 356 maybe could be finessed to say that *if* the underlying types have implementation-specific interface implementations then they are mapped to the non-alias contextual keywords… But that would only be needed until C#11 so maybe its a point not worth making?

So based on the referenced comment and the above para about line 356 this may be deleted (no change request as it has an embedded code block and I've no idea how to make a change suggestion including one of those!).

Author: This was removed.

>
> *end note*.

> Although `nint` and `nuint` shall be represented by the types `System.IntPtr` and `System.UIntPtr`, respectively, `nint` and `nuint` are *not* aliases for those types. As such, not all members of the corresponding `System` types are defined for `nint` and `nuint`. Instead, the compiler shall make available additional conversions, unary operators (§12.9) and arithmetic operators (§12.12) for the types `System.IntPtr` and `System.UIntPtr` when used in the context of native integer types.

Reviewer 3: We've said that there's an identity conversion between `nint` and `IntPtr` - but do we say elsewhere that `IntPtr` *must* be used to represent `nint` values? It may well be somewhere, and I've missed it...

Reviewer 3: (Line 365 of this file states it, effectively - but given that that is after this paragraph, presumably it's somewhere else as well...)

Reviewer 3: (Not sure what "shall be represented" even means in the language spec...)

Reviewer 2: I think we *might* be able to remove a lot of the distinction, but more work required for us to be sure. What I think we might be seeing here are two different things:

1. The *language* adding contextual keywords in C#9, and moving to keywords in C#11
2. An *implementation* changing how it provides the language features in its C#11 release compared to how it provided them in its C#9 & C#10 releases.

If this is correct then (1) is for the Standard and (2) is not.

Author: That implementation change likely has impact on the standard.

Until there was runtime support for `checked` and `unchecked` user defined operators, `System.IntPtr` and `System.UIntPtr` couldn't support the correct behavior for numeric types. We need to discuss how we want to handle that requirement.

While the language provides operations and conversions for `nint` and `nuint` that are appropriate for integer types, those operations and conversions are not available on the `System` type counterparts. For example,

<!-- Example: {template:"code-in-main-without-using", name:"SimpleTypes3", expectedException:"RuntimeBinderException"} -->
```csharp
M((nint)1);

static void M(dynamic d)
{
    var v = d >> 2; // RuntimeBinderException: '>>' cannot be applied to operands
                    // of type System.IntPtr/System.UIntPtr and int
}
```

The only constructor for `nint` or `nuint` is the parameter-less constructor.

The following members of `System.IntPtr` and `System.UIntPtr` are explicitly excluded from `nint` or `nuint`:

```csharp
// constructors
// arithmetic operators
// implicit and explicit conversions
public static readonly IntPtr Zero; // use 0 instead
public static int Size { get; }     // use sizeof() instead
public static IntPtr Add(IntPtr pointer, int offset);
public static IntPtr Subtract(IntPtr pointer, int offset);
public int ToInt32();
public long ToInt64();
public void* ToPointer();
```

> The remaining members of `System.IntPtr` and `System.UIntPtr` are implicitly included in `nint` and `nuint`. These are:

Reviewer 3: "These are" sounds like we're prohibiting the introduction of other members, which we've previously said would be okay for an implementation to do. Let's discuss what we want to require here.

Author: I removed the list. I think it's valid just to say "everything else"

```csharp
public override bool Equals(object obj);
public override int GetHashCode();
public override string ToString();
public string ToString(string format);
```

Interfaces implemented by `System.IntPtr` and `System.UIntPtr` are implicitly included in `nint` and `nuint`, with occurrences of the underlying types replaced by the corresponding native integer types. For example, if `IntPtr` implements `ISerializable`, `IEquatable<IntPtr>`, and `IComparable<IntPtr>`, then `nint` implements `ISerializable`, `IEquatable<nint>`, and `IComparable<nint>`.

`typeof(nint)` is `typeof(System.IntPtr)`, and `typeof(nuint)` is `typeof(System.UIntPtr)`.

> <!-- The following three paragraphs can be deleted for C# 10 -->

Reviewer 4: ```suggestion
<!-- The following three paragraphs can be deleted for C# 11 -->
```
